### PR TITLE
Validate existing suppressions and remove unnecessary suppressions when re-generating suppression files

### LINKED ISSUE #28070: Remove obsolete suppressions when re-generating suppression files
When an existing suppression file is re-generated, its content is read, deserialized into a set of suppressions and then stored in a list. When then at a later point, the suppressions are serialized to disk (because the ´--generate-suppression-file` flag is provided), the existing deserialized suppressions plus the new ones are written to disk. Therefore, obsolete suppressions which don't apply anymore are preserved and never get trimmed out.

It would be good to offer a flag to trim out such obsolete suppressions, similar to the one that the legacy APICompat tooling already offers: https://github.com/dotnet/arcade/blob/025103bcaefad81506465eeb7bb09b107b20f32d/src/Microsoft.DotNet.ApiCompat/src/ApiCompatRunner.cs#L45

### PR DESCRIPTION
**Pull request by ViktorHofer**
Fixes https://github.com/dotnet/sdk/issues/28070

Adds the following flags to the APICompat frontends:

| MSBuild | CLI | Default value |
| --- | --- | --- |
| /p:ApiCompatPreserveUnnecessarySuppressions | --preserve-unnecessary-suppressions | false |
| /p:ApiCompatPermitUnnecessarySuppressions | --permit-unnecessary-suppressions | false |

dotnet/runtime will need to opt-out of these switches for packable projects which have a reference assembly as we do a two phase validation for these.

**Note for reviewers: Hide whitespaces and exclude resx and xlf changes in the GH review pane.**

### CODE REVIEW

File: src/ApiCompat/Microsoft.DotNet.ApiCompat.Shared/SuppressionFileHelper.cs [hunk 1]
```diff
@@ -25,14 +26,14 @@ internal static class SuppressionFileHelper
 
             if (suppressionOutputFile == null)
             {
-                throw new ArgumentException(CommonResources.SuppressionsFileNotSpecified, nameof(suppressionOutputFile));
+                log.LogError(CommonResources.SuppressionsFileNotSpecified);
+                return;
```
Comment (ViktorHofer) [inline src/ApiCompat/Microsoft.DotNet.ApiCompat.Shared/SuppressionFileHelper.cs:30]: Unrelated change. Before, we were throwing an exception when a suppression file path wasn't specified but the generate suppression file property/flag was set. Now, with our improved logging interface, we can actually emit an error without a suppression (which wasn't possible before).

File: src/ApiCompat/Microsoft.DotNet.ApiCompat.Task/ValidatePackageTask.cs [hunk 2]
```diff
@@ -34,11 +34,6 @@ public class ValidatePackageTask : TaskBase
         /// </summary>
         public string? RuntimeGraph { get; set; }
 
-        /// <summary>
-        /// A NoWarn string contains the error codes that should be ignored.
-        /// </summary>
-        public string? NoWarn { get; set; }
-
```
Comment (ViktorHofer) [inline src/ApiCompat/Microsoft.DotNet.ApiCompat.Task/ValidatePackageTask.cs:41]: Unrelated change. Just moved the NoWarn down to mimic the other task and the CLI order.

File: src/ApiCompat/Microsoft.DotNet.ApiCompat.Tool/Program.cs [hunk 3]
```diff
@@ -42,7 +48,7 @@ static int Main(string[] args)
             };
             Option<MessageImportance> verbosityOption = new(new string[] { "--verbosity", "-v" },
                 "Controls the log level verbosity. Allowed values are high, normal, and low.");
-            verbosityOption.SetDefaultValue(MessageImportance.High);
+            verbosityOption.SetDefaultValue(MessageImportance.Normal);
```
Comment (ViktorHofer) [inline src/ApiCompat/Microsoft.DotNet.ApiCompat.Tool/Program.cs:51]: Unrelated change. I thought that we already downgraded the verbosity for the CLI, but apparently not. This should not be high as we want the summary confirmation message to be printed when invoking the CLI tool.

File: src/ApiCompat/Microsoft.DotNet.ApiCompat.Shared/CommonResources.resx [hunk 4]
```diff
@@ -126,6 +126,9 @@
   <data name="NoBreakingChangesFound" xml:space="preserve">
     <value>APICompat ran successfully without finding any breaking changes.</value>
   </data>
+  <data name="ObsoleteBaselineSuppressionsFound" xml:space="preserve">
+    <value>Obsolete suppressions found that should be removed from suppression files:</value>
```
Comment (ericstj) [inline src/ApiCompat/Microsoft.DotNet.ApiCompat.Shared/CommonResources.resx:130]: The term `obsolete` doesn't seem correct for these cases.  A term like `unnecessary` might be more clear.  That's the term which is used in https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0079.

File: src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/Suppression.cs [hunk 5]
```diff
@@ -79,5 +91,52 @@ public override int GetHashCode()
...
+            }
+
+            stringBuilder.Append(DiagnosticId);
+            stringBuilder.Append(" (");
+
+            bool requiresDelimiter = false;
+
+            if (Target is not null)
+            {
+                stringBuilder.AppendFormat("{0}: '{1}'", Resources.SuppressionTarget, Target);
```
Comment (ericstj) [inline src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/Suppression.cs:112]: Given the XML itself is not localized it might not make sense to localize this.

File: src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/SuppressionEngine.cs [hunk 6]
```diff
@@ -86,91 +93,61 @@ public bool IsErrorSuppressed(Suppression error)
...
+
+        protected virtual Stream GetWritableStream(string suppressionFile) => new FileStream(suppressionFile, FileMode.Create);
 
         private HashSet<Suppression> ParseSuppressionFiles(string[]? suppressionFiles)
         {
             HashSet<Suppression> suppressions = new();
 
             if (suppressionFiles != null)
             {
+                XmlSerializer _serializer = GetXmlSerializer();
```
Comment (ericstj) [inline src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/SuppressionEngine.cs:150]: don't prefix locals with `_`

File: src/ApiCompat/Microsoft.DotNet.ApiCompat.Tool/Program.cs [hunk 7]
```diff
@@ -190,6 +202,7 @@ static int Main(string[] args)
             runApiCompatOption.SetDefaultValue(true);
             Option<bool> enableStrictModeForCompatibleTfmsOption = new("--enable-strict-mode-for-compatible-tfms",
                 "Validates api compatibility in strict mode for contract and implementation assemblies for all compatible target frameworks.");
+            enableStrictModeForCompatibleTfmsOption.SetDefaultValue(true);
```
[the remark this answers is not in the document]
Comment (ericstj) [reply src/ApiCompat/Microsoft.DotNet.ApiCompat.Tool/Program.cs:205]: As I mentioned where you changed this in the task -- have another look -- I don't think this should default to true.

File: src/ApiCompat/Microsoft.DotNet.ApiCompat.Task/ValidatePackageTask.cs [hunk 8]
```diff
@@ -67,7 +62,7 @@ public class ValidatePackageTask : TaskBase
         /// <summary>
         /// Enables strict mode api comparison checks enqueued by the compatible tfm validator.
         /// </summary>
-        public bool EnableStrictModeForCompatibleTfms { get; set; }
+        public bool EnableStrictModeForCompatibleTfms { get; set; } = true;
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/ApiCompat/Microsoft.DotNet.ApiCompat.Task/ValidatePackageTask.cs:65]: That's already the default today, just in the entrypoint target instead of in the task: https://github.com/dotnet/sdk/blob/a7f5021d76002cfc1020715a0c777d9c5f0e7b72/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ApiCompat.ValidatePackage.targets#L48

[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/ApiCompat/Microsoft.DotNet.ApiCompat.Task/ValidatePackageTask.cs:65]: That was introduced with https://github.com/dotnet/sdk/pull/18714#discussion_r663448216

File: src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/SuppressionEngine.cs [hunk 9]
```diff
@@ -6,46 +6,39 @@
...
     /// </summary>
     public class SuppressionEngine : ISuppressionEngine
     {
-        public const string DiagnosticIdDocumentationComment = " https://learn.microsoft.com/en-us/dotnet/fundamentals/package-validation/diagnostic-ids ";
-        protected HashSet<Suppression> _validationSuppressions;
-        private readonly ReaderWriterLockSlim _readerWriterLock = new();
-        private readonly XmlSerializer _serializer = new(typeof(Suppression[]), new XmlRootAttribute("Suppressions"));
+        protected const string DiagnosticIdDocumentationComment = " https://learn.microsoft.com/en-us/dotnet/fundamentals/package-validation/diagnostic-ids ";
+        private readonly HashSet<Suppression> _baselineSuppressions;
+        private readonly HashSet<Suppression> _suppressions = new();
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/SuppressionEngine.cs:22]: The lock protected the suppressions HashSet before. With the introduction of another HashSet for the baseline suppressions, I would have needed another lock. As this code is never executed from more than one thread at the same time, the lock isn't necessary and just adds overhead. At this point of time, we don't need multi-threading support in the SuppressionEngine.

File: src/ApiCompat/Microsoft.DotNet.ApiCompat.Task/ValidatePackageTask.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/ApiCompat/Microsoft.DotNet.ApiCompat.Task/ValidatePackageTask.cs]: Reverted the change here and opened https://github.com/dotnet/sdk/issues/33011 to discuss.

File: src/ApiCompat/Microsoft.DotNet.ApiCompat.Task/ValidatePackageTask.cs [hunk 10]
```diff
@@ -89,6 +84,16 @@ public class ValidatePackageTask : TaskBase
...
 
+        /// <summary>
+        /// If true, removes obsolete baseline suppressions when re-generating the suppression file.
+        /// </summary>
+        public bool RemoveObsoleteSuppressions { get; set; } = true;
+
+        /// <summary>
+        /// If true, validates that baseline suppressions aren't obsolete.
+        /// </summary>
+        public bool ValidateSuppressions { get; set; } = true;
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/ApiCompat/Microsoft.DotNet.ApiCompat.Task/ValidatePackageTask.cs]: Every new rule in APICompat that is on by default is considered a breaking change. As well as infrastructure changes that enable more validation, i.e. https://github.com/dotnet/sdk/pull/32930. I will group both changes together into one breaking change notice.

File: src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/SuppressionEngine.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (ericstj) [reply src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/SuppressionEngine.cs:22]: > this code is never executed from more than one thread

I see, that must have changed at some point.

**Review by ericstj — COMMENTED** [review]
Changes look good.  Thank you for addressing.  I didn't see tests yet so I imagine that's still pending.

I left one small observation now that we've aligned the properties.

**Comment by ViktorHofer** [comment]
@ericstj can you please take another look? I just pushed two more commits which add tests, refactor the existing ones. I also found one bug in the WriteSuppressionFilesToFile method which I fixed by creating a new HashSet based on the existing one for the suppressions.

The tests were quite messy so I spent some time cleaning them up.

**Review by ericstj — APPROVED** [review]
LGTM - modulo one nit on the tests.  Thank you for adding them.

File: src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Logging/SuppressionEngineTests.cs [hunk 11]
```diff
@@ -12,120 +13,179 @@ namespace Microsoft.DotNet.ApiCompatibility.Logging.Tests
...
-
-            // Test IsErrorSuppressed Suppression overload.
-            Assert.True(testEngine.IsErrorSuppressed(new Suppression("CP0001")
-            {
-                Target = "T:A.B",
-                Left = "ref/netstandard2.0/tempValidation.dll",
-                Right = "lib/net6.0/tempValidation.dll"
-            }));
+            TestSuppressionEngine suppressionEngine = new();
+            suppressionEngine.LoadSuppressions("NonExistentFile.xml");
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Logging/SuppressionEngineTests.cs:31]: Yes, that's essential as it trigger loading the test suppression file content. See TestSuppressionEngine.cs which has an override to get the readable stream. Without calling the newly introduced `LoadSuppressions` method, the baseline suppressions won't get loaded.

**Comment by RussKie** [comment]
Thank you :)
